crates/pathfinder/src/state/merkle_tree.rs

```diff
@@ -81,6 +82,64 @@ pub trait NodeStorage {
     fn increment_ref_count(&self, key: StarkHash) -> anyhow::Result<()>;
 }
 
+/// Lightweight representation of [BinaryNode]. Only holds left and right hashes.
+#[allow(dead_code)]
+#[derive(Debug, PartialEq, Serialize)]
+pub struct BinaryProofNode {
+    pub left_hash: StarkHash,
+    pub right_hash: StarkHash,
+}
+
+impl From<&BinaryNode> for BinaryProofNode {
+    fn from(bin: &BinaryNode) -> Self {
+        Self {
+            left_hash: bin.left.borrow().hash().expect("Node should be committed"),
+            right_hash: bin.right.borrow().hash().expect("Node should be committed"),
+        }
+    }
+}
+
+/// Ligthtweight representation of [EdgeNode]. Only holds its path and its child's hash.
+#[allow(dead_code)]
+#[derive(Debug, PartialEq, Serialize)]
+pub struct EdgeProofNode {
+    #[serde(skip)]
+    pub path: BitVec<Msb0, u8>,
+    pub child_hash: StarkHash,
+}
+
+impl From<&EdgeNode> for EdgeProofNode {
+    fn from(edge: &EdgeNode) -> Self {
+        Self {
+            path: edge.path.clone(),
+            child_hash: edge
+                .child
+                .borrow()
+                .hash()
+                .expect("Node should be committed"),
+        }
+    }
+}
+
+/// [ProofNode] s are lightweight versions of their `Node` counterpart.
+/// They only consist of [BinaryProofNode] and [EdgeProofNode] because `Leaf`
+/// and `Unresolved` nodes should not appear in a proof.
+#[derive(Debug, PartialEq, Serialize)]
+pub enum ProofNode {
+    Binary(BinaryProofNode),
+    Edge(EdgeProofNode),
+}
+
+impl From<&Node> for ProofNode {
+    fn from(node: &Node) -> Self {
+        match node {
+            Node::Binary(bin) => Self::Binary(BinaryProofNode::from(bin)),
+            Node::Edge(edge) => Self::Edge(EdgeProofNode::from(edge)),
+            _ => unreachable!(),
+        }
+    }
+}
```

Comment:
Ok I get it. So you'd the caller having to write the `match` arms rather than it being "hidden" away?

My initial implem had a `try_from` trait implem (result in an Err instead of unreachable), but went for `from` for simplicity (because I knew that the invariant would hold in my code logic).

Should I go back to `try_from`, or do you think having the caller do the `match` logic is better?

Reply:
see 8c2679adca08abea337cf51c880c7846cea51478